lib/bap_disasm/bap_disasm_rec.ml

```diff
@@ -241,34 +157,88 @@ let next dis s =
     | [] -> Dis.stop dis s
     | r :: roots when not(Memory.contains s.base r) ->
       loop {s with roots}
-    | r :: roots when Span.mem s.visited r ->
+    | r :: roots when Visited.is_insn s.visited r ->
```

Comment:
Since now you visited set only records valid starts of instructions, not the set of addresses that were attempted by a disassembler, we may not now guarantee, that the recursive disassembler will terminate. A particular example (valid for the arm diassembler which prefers to skip through bytes, is the following):

inputs: `data=01234, root=[0]`
disassembler-output: `insn=1234, next=0`

Here we will mark `1234` as a valid instruction and visited, and continue with the same `root=[0]`, hence the infinite loop.

So we need to either check every time when we add a new root that it wasn't attempted, or check the same property every time we extract a root.